### PR TITLE
XfodDvJh: Unify version definitions

### PR DESCRIPTION
**Pull request by omikron-fp**
Introduce build step at Node-level that copies the version from `package.json` into the JS output.
This reduces the number of locations where the version is defined from two to one.

### CODE REVIEW

**Comment by loop-evgeny** [comment]
Looks like this is only for unifying version definitions, but that's a good thing. Mixing unrelated changes into one PR just makes it harder to review, and harder to revert f needed. Please make the user ID placeholder a separate PR (even if it's a very small change).

File: README.md [hunk 1]
```diff
@@ -179,6 +179,10 @@ A Linux or MacOS environment is required. Windows is not supported and the npm c
 3. run `npm run dev` to start the webserver, open up
    http://localhost:3001/#search and try out the basic features
 
+Before merging a branch, create a commit that updates the package's version number in `package.json`.
```
Comment (loop-evgeny) [inline README.md:182]: I don't quite understand - does the version change need to be a separate commit for some reason? We usually just update the same version in the same PR that makes other changes, which may be 1 or more commits.

File: README.md [hunk 2]
```diff
@@ -179,6 +179,10 @@ A Linux or MacOS environment is required. Windows is not supported and the npm c
...
+Also, update the version in `package-lock.json`, either manually or through NPM (e.g. `npm i`).
```
Comment (loop-evgeny) [inline README.md:183]: Why would package-lock.json ever need to be updated manually? I thought that's updated automatically by the build process?

Comment (loop-evgeny) [reply README.md:183]: What happens if someone forgets this step and the versions in package.json and package-lock.json differ?

File: replaceVersion.js [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+
+
+process.stdin.on(`data`, input => {
+    chunks.push(input.toString());
+});
+
+process.stdin.on(`end`, () => {
+    const replaced = chunks.join(``).replaceAll(`___REPLACED DURING BUILD JOB___`, version);
+
+    console.log(replaced);
```
Comment (loop-evgeny) [inline replaceVersion.js:13]: Doesn't this add a newline? Why not `process.stdout.write`?

File: src/core.js [hunk 4]
```diff
@@ -4,7 +4,7 @@ import cookies from "./cookies.js";
 let core = {
 
     versions: {
-        libVersion: "1.12.5454545454-build-number", //"5454545454-build-number" will be replaced by teamcity. also in package.json
+        libVersion: "___REPLACED DURING BUILD JOB___",
```
Comment (loop-evgeny) [inline src/core.js:7]: This could well be a silly question, revealing my ignorance of JS, but is it possible for someone to run the connector without "building" it (like during development), and thus actually send this placeholder as the `Lib-Version` header? Maybe it's safer go the other way and have the version copied from here to package.json?

**Review by loop-evgeny — COMMENTED** [review]
You probably already saw, but the TC build failed: https://teamcity.loop54.com/viewLog.html?buildId=175213&buildTypeId=Misc_Libraries_JsConnectorDev&tab=buildLog

Please go to https://teamcity.loop54.com/viewModification.html?modId=159594&personal=false&buildTypeId=Misc_Libraries_JsConnectorDev&tab=vcsModificationFiles and click the "it's me!" link, so TC associates your git username with your TC username and notifies you of build failures containing your commits.

**Comment by omikron-fp** [comment]
> Looks like this is only for unifying version definitions, but that's a good thing. Mixing unrelated changes into one PR just makes it harder to review, and harder to revert f needed. Please make the user ID placeholder a separate PR (even if it's a very small change).

PR wasn't complete yet.
But sure, I reduced the description to single-issue.

File: README.md (hunk 1 above)
Comment (omikron-fp) [reply README.md:182]: Committing the version separately is convenient when you need to cherry-pick implementation. Then you don't have to deal with unrelated changes that likely produce merge conflicts.
Same for README updates.

Would you like me to rephrase this?

File: README.md (hunk 2 above)
Comment (omikron-fp) [reply README.md:183]: Changes by the TC build pipeline are not persisted in the repo. If you don't care about an out of sync package-lock, we can ignore this step. It could be a minor source of confusion to somebody looking at package and package-lock files, but there no technical consequences.

File: src/core.js (hunk 4 above)
Comment (omikron-fp) [reply src/core.js:7]: In a Node.js setting this is possible.
Previously, this kind of usage would have sent `1.12.5454545454-build-number`. Is this less of an issue?

Generally speaking, I think `package.json` should be the source of the version number.
It is possible to dynamically read the version from `package.json` at runtime.
```js
{ libVersion: require("package.json").version }
```
For a Node bundle this might be fine.
But for a browser bundle this would have to get inlined. I tried this, but our current bundler simply puts the entire `package.json` into the final bundle and reads from it like a module. We can't do that because it wastes space and reveals the libraries dependencies and their versions.

Various solutions thinkable here. Which one do you prefer?

File: package.json [hunk 5]
```diff
@@ -5,7 +5,7 @@
   "license": "BSD-3-Clause",
   "main": "lib/loop54-js-connector-client.js",
   "scripts": {
-    "clean": "rm -rf ./lib",
+    "clean": "rm -rf ./lib && mkdir lib",
```
Comment (omikron-fp) [inline package.json:8]: Once reason the TC job failed was that the `/lib` directory is deleted and the `node replaceVersion.js > lib/loop54-js-connector.js` part of the script can't write a file in a directory that doesn't exist.

`npm run bundle` should be fine. But the piping at the script level now means that running `npm run build-connector` individually would fail if your don't have the `/lib` directory on your disk.

How much of an issue do you think is this?

File: replaceVersion.js (hunk 3 above)
Comment (omikron-fp) [reply replaceVersion.js]: Adjusted in fixup commit.

File: README.md (hunk 1 above)
Comment (loop-evgeny) [reply README.md:182]: Something like this would make it clearer to me:

"Before merging to master update the version in `package.json`, preferably in a separate commit."

What do you think?

File: README.md (hunk 2 above)
Comment (loop-evgeny) [reply README.md:183]: So TC runs `npm ci`, which shouldn't update package-lock.json anyway, as I understand. So I guess we could say something like

> Before creating a PR:
> - update the version in `package.json`, preferably in a separate commit
> - run `npm i` to update `package-lock.json` and commit the changes

But then do we want a separate commit for the `package-lock.json` too? Those 2 updates can be done in one commit, right? I.e. you do all your "substantial" changes, commit, update version, `npm i`, commit?

File: README.md (hunk 1 above)
Comment (omikron-fp) [reply README.md:182]: Works for me.
Perhaps we continue the conversation in the other thread, about `package-lock` at
https://github.com/LoopFiftyFour/JS-Connector/pull/53#discussion_r1778384028

File: README.md (hunk 2 above)
Comment (omikron-fp) [reply README.md:183]: Yes, the version updates should go into the same commit. The steps you mention are exactly what I am thinking.

For the current workload this workflow should work. But once there are multiple parallel PRs, they will compete for the version update. It also blocks publishing multiple changes in a single version increment.
Maybe a future topic to tackle.

File: src/core.js (hunk 4 above)
Comment (loop-evgeny) [reply src/core.js:7]: I think receiving `1.12.5454545454-build-number` as `Lib-Version` is not as bad, because then we still basically know the source code version.

OK, so if you want to have the version in package.json, and it's too hard to read it from package.json, then another option is to have it in both places (as before), but have the build fail if the versions differ. Maybe it's not so bad to have to update the version number in two places, as long as we ensure they stays in sync?

File: package.json (hunk 5 above)
Comment (loop-evgeny) [reply package.json:8]: I don't know, but can't it easily be fixed by adding the `mkdir lib` to `build-connector` instead? Seems strange to have the clean target re-create the output dir.

File: README.md (hunk 2 above)
Comment (loop-evgeny) [reply README.md:183]: Yes, I think that's fine. In the rare case of multiple parallel PRs we can just have only one of them update the version number, and that one is the last to be merged. Merging the other PRs (without a version number update) will simply fail to publish the npm package, since the version already exists, right?

File: src/core.js (hunk 4 above)
Comment (loop-evgeny) [reply src/core.js:7]: Sorry for the late reply. I don't think we need too much automation, given how rarely we actually update the JS connector. Personally I'd be happy with manually updating the version in two places, as long as we somehow prevent the two from getting out of sync. But if you're going to be the one maintaining this then it's really up to you!

**Comment by omikron-fp** [comment]
Rolled back most of the changes and went for a different approach.
PR needs squashing and commit message adjustments before merging.

File: syncVersion.js [hunk 6]
```diff
@@ -0,0 +1,11 @@
+const fs = require(`fs`);
+const { version } = require(`./package.json`);
+
+const coreFile = `./src/core.js`;
+
+
+const replacedCore = fs
+    .readFileSync(coreFile, `utf8`)
+    .replaceAll(/\d+\.\d+\.5454545454-build-number/g, version);
```
Comment (loop-evgeny) [inline syncVersion.js:9]: If this regex fails to match for some reason (like maybe the placeholder gets changed or the version gets moved to another file) then the version will silently remain out of sync. It would be good to detect that and fail here.

File: bundle.js [hunk 7]
```diff
@@ -3,16 +3,16 @@ const runAll = require("npm-run-all");
 const runopts = {
     parallel : true,
     stdout   : process.stdout,
+    stderr   : process.stderr,
```
Comment (omikron-fp) [inline bundle.js:6]: Had to add this to make the errors from `verifyVersion.js` visible.

**Comment by omikron-fp** [comment]
Okay. Last attempt hopefully.
As TeamCity replaces the build number placeholders before running the build script, the version verification script (which is run by the build script) has to detect whether the build number has been replaced already. This complicates things a bit and `verifyVersion.js` looks accordingly. But at least pretty much all possible scenarios of out-of-sync versions should be covered.

**Comment by loop-evgeny** [comment]
> Okay. Last attempt hopefully. As TeamCity replaces the build number placeholders before running the build script, the version verification script (which is run by the build script) has to detect whether the build number has been replaced already. This complicates things a bit and `verifyVersion.js` looks accordingly. But at least pretty much all possible scenarios of out-of-sync versions should be covered.

Couldn't we simplify things by having TeamCity only replace the build number in package.json? Then replacing the version in core.js would proceed the same way when running in TeamCity as when running normally, and should only require one regex.

**Comment by omikron-fp** [comment]
TC pipeline is changed to only update `package.json`.
It also got a separate build command `build-ci` because the `npm i` of the regular command is redundant to the CI pipeline.

Had to move the verification from the `build-(connector|client)` commands to the `bundle` command. The challenge with TC replacing the version placeholder before starting the build remains. In effect, the TC pipeline only replaces the version without verifying it again (which would be redundant anyway). The regular `bundle` command which is intended for local usage _does_ perform the verification.

File: README.md [hunk 8]
```diff
@@ -179,6 +179,19 @@ A Linux or MacOS environment is required. Windows is not supported and the npm c
 3. run `npm run dev` to start the webserver, open up
    http://localhost:3001/#search and try out the basic features
 
+### Pull requests
+
+Before creating a pull request:
+- commit your changes
+- update the package's version number in `package.json`
+- run `npm run sync-version`
```
[the remark this answers is not in the document]
Comment (loop-evgeny) [reply README.md:187]: Tested this now by committing a version change to package.json only. The build succeeds, and the versions are in sync in the output produced by TC, which is certainly better than what we had before, but they remain out of sync in source control.

It would be nice to have the TC build fail in that case, so the developer is immediately informed that they forgot to sync the versions and the PR with out-of-sync versions does not get merged. Can verifyVersion.js --sync still check that the part of the version before the build number (i.e. the major+minor version) is the same?

(Alternatively, we could have that script set the build number instead of using TC's text replacement "build feature", but I think that would get more complicated than checking the first 2 parts of the version.)

[the remark this answers is not in the document]
Comment (loop-evgeny) [reply README.md:187]: Edited my comment above because I realized the solution I was proposing was quite silly and would never work.

So maybe the version you had before (with two regexes) was the best solution, after all! Sorry about that.

File: bundle.js [hunk 9]
```diff
@@ -3,16 +3,16 @@ const runAll = require("npm-run-all");
...
         console.log("Running build");
-        runAll(["build-*"], runopts)
-            .then(() => {
-                console.log("Running uglify");
-                runAll(["uglify"], runopts)
-            })
+        return runAll(["build-*"], runopts);
     })
-    .catch(console.error);
+    .then(() => {
```
[the remark this answers is not in the document]
Comment (omikron-fp) [reply bundle.js:15]: It does change the meaning in that the Promises are now properly chained up. The `catch` was attached only to `runAll(["clean"], runopts)`. Exceptions thrown by the nested Promises were uncaught.

I indeed removed the `catch` to have the CI build fail.

File: README.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (omikron-fp) [reply README.md:187]: I think I finally understand where you want to go with this topic. Let me analyse our options and constraints.

The challenge we are facing is that the version verification must be a Node script that relies on placeholders (or special comments, which suck). And this script can only be run as a TC build step.
At the same time, TC must inject the build number in place of the placeholder and this can only happen in the content replacer feature which runs **before** we can run any Node script.

**Side question**: What is the purpose of the build number? Do we ever publish a minor version twice but with a different build number? E.g. `1.14.178` then `1.14.180`.

So if we want the build number, we **must** use the content replacer build feature.
If we want to catch out of sync versions in source control, we **must** use a Node script with placeholders intact.

Here is what I think would work:

`package.json` and `core.js` keep their placeholders as they are now.
We introduce a new constant with a separate placeholder in `verifyVersion.js` that gets replaced with TC's build number.
```js
const buildNumber = `TC-BUILD-NUMBER`;
```
TC only replaces this new placeholder and no longer `package` and `core`.
This way we have the build number in the code but the version placeholders are still intact.
TC triggers build step _verify version_. Here we have the state from version control. Fails if out of sync.
TC triggers _sync version_. Node script replaces placeholders in `package` and `core` with the newly introduced `buildNumber` from above.
Build.

**Locally**, we can keep the procedure of manually updating `package.json` then running _sync version_.
It's likely necessary to have separate sync scripts for local and CI build because locally, the build number is not known.
Local and CI build scripts also must likely remain separate as they currently are because of the differing placeholder availability.

Let me know your thoughts.

[the remark this answers is not in the document]
Comment (loop-evgeny) [reply README.md:187]: > TC must inject the build number in place of the placeholder and this can only happen in the content replacer feature which runs before we can run any Node script.

I don't think it has to be there - we can inject the build number as part of a Node script, too, it was just simpler to use the built-in replacer feature for this.

But I think you ask a great question about whether the build number is needed! I don't know, but it's possible that we simply copied what we do for the engine, where we do have many builds of the same major/minor version and it's good to know the _exact_ version.

We can of course make multiple builds with the same major/minor version of the JS Connector, too, and [we have in the past](https://www.npmjs.com/package/loop54-js-connector?activeTab=versions), but I don't think we should - it's confusing to customers and inconsistent with what we do for the .NET Connector. So to prevent that it might actually be good to not have the build number in package.json (but keep it in core.js).

Then we can still do the replacement using the TC content replacer feature, as today. Extracting the build number into its own constant is still helpful, I think: then we can just verify that `"version in core.js" = "version in package.json" + buildNumber.`, so that check becomes independent of the build number.

**Comment by omikron-fp** [comment]
All version related operations are now in `versioning.js`.

I amended the TC pipeline to always verify the versions in the source as its first step.
It injects the build number into a designated JS constant without touching the placeholders in the actual version numbers.
TC and local version verification now use the same script.

Local build script includes version verification. TC build script does not, because it happened earlier in the pipeline and the placeholder will have been overwritten already.

**Review by loop-evgeny — APPROVED** [review]
Very nice! Tested with and without syncing the version numbers.

[With](https://teamcity.loop54.com/viewLog.html?buildId=176824&buildTypeId=Misc_Libraries_JsConnectorDev&tab=buildLog&_focus=136#_state=123,136):

> [Step 1/10] Matched version: 1.98.5454545454-build-number
> [Step 1/10] Version successfully verified.
> [Step 2/10] Injected build number: 179

[Without](https://teamcity.loop54.com/viewLog.html?buildId=176825&buildTypeId=Misc_Libraries_JsConnectorDev&tab=buildLog&branch_Misc_Libraries=Qg21A1yi-test-version2):

> [Step 1/10] Error: Versions are out of sync. See README about version syncing.
> [Step 1/10] ./package.json: 1.99.5454545454-build-number
> [Step 1/10] ./src/core.js: 1.98.5454545454-build-number
